### PR TITLE
Don't count up key presses during breaks

### LINKED ISSUE #2707: Key presses during break times count up towards total
They do not on stable

### PR DESCRIPTION
**Pull request by HoutarouOreki**
Resolves #2707 
Hopefully the code isn't too bad

---

### CODE REVIEW

**Review by swoolcock — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/BreakOverlay.cs [hunk 1]
```diff
@@ -120,6 +126,9 @@ private void initializeBreaks()
                 if (!b.HasEffect)
                     continue;
 
+                Scheduler.AddDelayed(OnBreakEnter, b.StartTime);
+                Scheduler.AddDelayed(OnBreakLeave, b.StartTime + b.Duration);
```
[the remark this answers is not in the document]
Comment (HoutarouOreki) [reply osu.Game/Screens/Play/BreakOverlay.cs]: It's pause proof, but not rewind resistant.

File: osu.Game/Screens/Play/BreakOverlay.cs [hunk 2]
```diff
@@ -40,6 +41,11 @@ public List<BreakPeriod> Breaks
         private readonly BreakInfo info;
         private readonly BreakArrows breakArrows;
 
+        public Action OnBreakEnter;
+        public Action OnBreakLeave;
+
+        public bool IsBreakTime;
```
[the remark this answers is not in the document]
Comment (swoolcock) [reply osu.Game/Screens/Play/BreakOverlay.cs]: On second thought, this doesn't really belong here at all.  `BreakOverlay` should be purely visual.  We should be calculating this in another class based on the current time and the `BreakPeriod`s provided for the beatmap.

[the remark this answers is not in the document]
Comment (HoutarouOreki) [reply osu.Game/Screens/Play/BreakOverlay.cs]: I can't disagree with this

File: osu.Game/Screens/Play/BreakOverlay.cs [hunk 3]
```diff
@@ -40,6 +41,11 @@ public List<BreakPeriod> Breaks
         private readonly BreakInfo info;
         private readonly BreakArrows breakArrows;
 
+        public Action OnBreakEnter;
+        public Action OnBreakLeave;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/BreakOverlay.cs]: How does events improve it though? You can still unassign events.

[the remark this answers is not in the document]
Comment (swoolcock) [reply osu.Game/Screens/Play/BreakOverlay.cs]: You can't unassign an event handler if you don't have a reference to it though, right?  If the event handler method is private then external classes shouldn't be able to remove it.  Regardless, I don't believe these actions belong in this class anyway.  They're related to the player position within `BreakPeriod`s, and nothing to do with the visual representation.

File: osu.Game/Screens/Play/BreakOverlay.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/BreakOverlay.cs]: we can't use schedules here. we only recently got rid of previously placed schedules in this class which were breaking rewind.

**Comment by HoutarouOreki** [comment]
Will try to think about a more proper solution. If anyone wants, feel free to take over that issue though.

**Pull request closed without merging**